src/database/postgres.go

```diff
@@ -0,0 +1,198 @@
+            )
+            VALUES($1, $2, $3, $4, $5, $6, $7, to_tsvector('english', $7), $8)
+            ON CONFLICT (name) DO UPDATE SET
+                uri = EXCLUDED.uri,
+                data_type = EXCLUDED.data_type,
+                source_type = EXCLUDED.source_type,
+                last_refreshed_time = EXCLUDED.last_refreshed_time,
+                content = EXCLUDED.content,
+                search_vector = to_tsvector('english', EXCLUDED.content),
+                word_occurrences = EXCLUDED.word_occurrences
+        `, entry.Name, entry.URI, entry.DataType.String(), entry.SourceType.String(),
+            entry.FirstAddedTime.AsTime(), entry.LastRefreshedTime.AsTime(),
+            entry.Content, wordOccurrencesJSON)
+    }
+
+    br := tx.SendBatch(ctx, batch)
+    _, err = br.Exec()
+    br.Close()
+    if err != nil {
+        return fmt.Errorf("failed to insert rows: %w", err)
+    }
+
+    err = tx.Commit(ctx)
+    if err != nil {
+        return fmt.Errorf("failed to commit transaction: %w", err)
+    }
+
+    // Refresh the materialized view outside the transaction
+    _, err = conn.Exec(ctx, `REFRESH MATERIALIZED VIEW CONCURRENTLY mv_index_list`)
+    if err != nil {
+        return fmt.Errorf("failed to refresh materialized view: %w", err)
+    }
+
+    return nil
+}
+
+
+// queryRow retrieves a single row from the index_list table based on the provided name.
+// It returns a pointer to a pb.IndexListEntry struct containing the row data, or an error if the query fails.
+//
+// Parameters:
+//   - ctx: The context for database operations.
+//   - conn: A PgxIface interface for database connection.
+//   - name: The name of the index entry to retrieve.
+//
+// Returns:
+//   - *pb.IndexListEntry: A pointer to the retrieved index entry.
+func queryRow(ctx context.Context, conn PgxIface, name string) (*pb.IndexListEntry, error) {
+	tx, err := conn.Begin(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to database: %w", err)
+	}
+	defer tx.Rollback(ctx)
+
+	var entry pb.IndexListEntry
+	var dataType, sourceType string
+	var firstAddedTime, lastRefreshedTime time.Time
+	var wordOccurrencesJSON []byte
+
+	err = tx.QueryRow(ctx, `
+		SELECT name, uri, data_type, source_type, first_added_time, last_refreshed_time, content, word_occurrences
+		FROM index_list 
+		WHERE name=$1
+	`, name).Scan(
+		&entry.Name, &entry.URI, &dataType, &sourceType, &firstAddedTime, &lastRefreshedTime, &entry.Content, &wordOccurrencesJSON)
+
+	if err != nil {
+		if err == pgx.ErrNoRows {
+			return nil, fmt.Errorf("no entry found for name: %s", name)
+		}
+
+		return nil, fmt.Errorf("QueryRow failed: %w", err)
+	}
+
+	entry.DataType = pb.DataType(pb.DataType_value[dataType])
+	entry.SourceType = pb.SourceType(pb.SourceType_value[sourceType])
+	entry.FirstAddedTime = timestamppb.New(firstAddedTime)
+	entry.LastRefreshedTime = timestamppb.New(lastRefreshedTime)
+
+	err = json.Unmarshal(wordOccurrencesJSON, &entry.WordOccurrences)
```

Comment:
It seems really expensive to return both the entire wordOccurrences map and the entire content when in most cases we'd only really want to get the URI, SourceType, and DataType back for the core retrieval use case. Can you change this to return those types only and rename it to something like getContentMetadata? We will probably need to handle queries involving the word occurrences map and other fields in other functions

Reply:
Yeah we can make more functions for them.